src/stream/src/executor/top_n.rs

```diff
@@ -91,29 +155,80 @@ pub struct InnerTopNExecutorNew<S: StateStore> {
 
 const TOPN_CACHE_HIGH_CAPACITY_FACTOR: usize = 2;
 
-pub struct TopNCache {
+pub struct TopNCache<const WITH_TIES: bool> {
     /// Rows in the range `[0, offset)`
     pub low: BTreeMap<OrderedRow, Row>,
     /// Rows in the range `[offset, offset+limit)`
+    ///
+    /// When `WITH_TIES` is true, it also stores ties for the last element,
+    /// and thus the size can be larger than `limit`.
     pub middle: BTreeMap<OrderedRow, Row>,
     /// Rows in the range `[offset+limit, offset+limit+high_capacity)`
+    ///
+    /// When `WITH_TIES` is true, it also stores ties for the last element,
+    /// and thus the size can be larger than `high_capacity`.
     pub high: BTreeMap<OrderedRow, Row>,
```

Comment:
If I understand correctly, the `high` region (i.e. the elements skipped by `offset`) will also be overflowed, right? As far as I've seen, This doesn't seem to be a supposed behavior of `WITH TIES`.

Thus, maybe we should comment that the generic variable `WITH_TIES` is only designed for `rank()`, or just rename it into something else.

Reply:
You mean the `low` region? And yes, as disscussed with @TennyZhuang, I simply banned `offset` now. Maybe decide later.

```rust
pub fn new(offset: usize, limit: usize, sort_key_len: usize) -> Self {
    assert!(limit != 0);
    if WITH_TIES {
        // It's trickier to support.
        // Also `OFFSET WITH TIES` has different semantic with `a < RANK() < b`
        assert!(offset == 0, "OFFSET is not supported with WITH TIES");
    }
```

Reply:
No, I mean `high` 🤣 What I said is, this behavior seems only to work with the range of `rank()` e.g. `10 < rank() < 20`. I don't think it's expected behavior for 

```
SELECT * from t 
order by salary desc 
LIMIT 3 OFFSET 5 WITH TIES
```

(because `OFFSET 5` is not actually supported here? I guess)

Reply:
low is skipped items and high is cache for later items. Why the high‘s behavior will be different？🤨